File: types/blob_test.go
package types

import (
	"bytes"
	"io"
	"testing"

	"github.com/attic-labs/noms/ref"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
)

// IMPORTANT: These tests and in particular the hash of the values should stay in sync with the
// corresponding tests in js

type countingReader struct {
	last uint32
	val  uint32
	bc   uint8
}

func newCountingReader() *countingReader {
	return &countingReader{0, 0, 4}
}

func (rr *countingReader) next() byte {
	if rr.bc == 0 {
		rr.last = rr.last + 1
		rr.val = rr.last
		rr.bc = 4
	}

	retval := byte(uint64(rr.val) & 0xff)
	rr.bc--
	rr.val = rr.val >> 8
	return retval
}

func (rr *countingReader) Read(p []byte) (n int, err error) {
	for i := 0; i < len(p); i++ {
		p[i] = rr.next()
	}
	return len(p), nil
}

func randomBuff(powOfTwo uint) []byte {
	length := 1 << powOfTwo
	rr := newCountingReader()
	buff := make([]byte, length)
	rr.Read(buff)
	return buff
}

type blobTestSuite struct {
	suite.Suite
	blob                   Blob
	buff                   []byte
	expectRef              ref.Ref
	expectChunkCount       int
	expectPrependChunkDiff int
	expectAppendChunkDiff  int
}

func newBlobTestSuite(size uint, expectRefStr string, expectChunkCount int, expectPrependChunkDiff int, expectAppendChunkDiff int) *blobTestSuite {
	buff := randomBuff(size)
	blob := NewBlob(bytes.NewReader(buff))
	return &blobTestSuite{
		blob:                   blob,
		buff:                   buff,
		expectRef:              ref.Parse(expectRefStr),
		expectChunkCount:       expectChunkCount,
		expectPrependChunkDiff: expectPrependChunkDiff,
		expectAppendChunkDiff:  expectAppendChunkDiff,
	}
}

func TestBlobSuite1K(t *testing.T) {
	suite.Run(t, newBlobTestSuite(10, "sha1-a8a3656441ac61edd9cb734559a6c081b5268a6d", 3, 2, 2))
}

func TestBlobSuite4K(t *testing.T) {
	suite.Run(t, newBlobTestSuite(12, "sha1-7fef6ea1ab709b1ea06446976d7fce02be164398", 9, 2, 2))
}

func TestBlobSuite16K(t *testing.T) {
	suite.Run(t, newBlobTestSuite(14, "sha1-16a7098cc077def534932f5a566d0b0ef60283b5", 33, 2, 2))
}

func TestBlobSuite64K(t *testing.T) {
	suite.Run(t, newBlobTestSuite(16, "sha1-a8c6563067edd1c0a49c8b30a2c75cc9d53c0fd3", 4, 2, 2))
}

func TestBlobSuite256K(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping test in short mode.")
	}
	suite.Run(t, newBlobTestSuite(18, "sha1-b924c4a13eb0542766a93f4329f4d13315ef823d", 13, 2, 2))
}

func (suite *blobTestSuite) TestRef() {
	suite.Equal(suite.expectRef.String(), suite.blob.Ref().String())
}

func (suite *blobTestSuite) TestType() {
	suite.True(BlobType.Equals(suite.blob.Type()))
}

func (suite *blobTestSuite) TestLen() {
	suite.Equal(len(suite.buff), int(suite.blob.Len()))
}

func (suite *blobTestSuite) TestEquals() {
	b2 := suite.blob
	suite.True(suite.blob.Equals(b2))
	suite.True(b2.Equals(suite.blob))
}

func (suite *blobTestSuite) TestChunkCount() {
	suite.Equal(suite.expectChunkCount, len(suite.blob.Chunks()))
}

func (suite *blobTestSuite) TestChunkRefType() {
	for _, r := range suite.blob.Chunks() {
		suite.True(RefOfBlobType.Equals(r.Type()))
	}
}

func (suite *blobTestSuite) TestRoundTripAndReadFull() {
	vs := NewTestValueStore()
	r := vs.WriteValue(suite.blob)
	v2 := vs.ReadValue(r.TargetRef()).(Blob)
	suite.True(v2.Equals(suite.blob))
	out := make([]byte, len(suite.buff))
	n, err := io.ReadFull(v2.Reader(), out)
	suite.NoError(err)
	suite.Equal(len(suite.buff), n)
	suite.Equal(suite.buff, out)
}

// Checks the first 1/2 of the bytes, then 1/2 of the remainder, then 1/2 of the remainder, etc...
func (suite *blobTestSuite) TestRandomRead() {
	buffReader := bytes.NewReader(suite.buff)
	blobReader := suite.blob.Reader()

	readByteRange := func(r io.ReadSeeker, start int64, count int64) []byte {
		bytes := make([]byte, count)
		n, err := r.Seek(start, 0)
		suite.NoError(err)
		suite.Equal(start, n)
		n2, err := io.ReadFull(r, bytes)
		suite.NoError(err)
		suite.Equal(int(count), n2)
		return bytes
	}

	checkByteRange := func(start int64, count int64) {
		expect := readByteRange(buffReader, start, count)
		actual := readByteRange(blobReader, start, count)
		suite.Equal(expect, actual)
	}

	length := int64(len(suite.buff))
	start := int64(0)
	count := int64(length / 2)
	for count > 2 {
		checkByteRange(start, count)
		start = start + count
		count = (length - start) / 2
	}
}

func chunkDiffCount(c1 []Ref, c2 []Ref) int {
	count := 0
	refs := make(map[ref.Ref]int)

	for _, r := range c1 {
		refs[r.TargetRef()]++
	}

	for _, r := range c2 {
		if c, ok := refs[r.TargetRef()]; ok {
			if c == 1 {
				delete(refs, r.TargetRef())
			} else {
				refs[r.TargetRef()] = c - 1
			}
		} else {
			count++
		}
	}

	count += len(refs)
	return count
}

func (suite *blobTestSuite) TestPrependChunkDiff() {
	dup := make([]byte, len(suite.buff)+1)
	dup[0] = 0
	copy(dup[1:], suite.buff)
	b2 := NewBlob(bytes.NewReader(dup))
	suite.Equal(suite.expectPrependChunkDiff, chunkDiffCount(suite.blob.Chunks(), b2.Chunks()))
}

func (suite *blobTestSuite) TestAppendChunkDiff() {
	dup := make([]byte, len(suite.buff)+1)
	copy(dup, suite.buff)
	dup[len(dup)-1] = 0
	b2 := NewBlob(bytes.NewReader(dup))
	suite.Equal(suite.expectAppendChunkDiff, chunkDiffCount(suite.blob.Chunks(), b2.Chunks()))
}

type testReader struct {
	readCount int
	buf       *bytes.Buffer
}

func (r *testReader) Read(p []byte) (n int, err error) {
	r.readCount++

	switch r.readCount {
	case 1:
		for i := 0; i < len(p); i++ {
			p[i] = 0x01
		}
		io.Copy(r.buf, bytes.NewReader(p))
		return len(p), nil
	case 2:
		p[0] = 0x02
		r.buf.WriteByte(p[0])
		return 1, io.EOF
	default:
		return 0, io.EOF
	}
}

func TestBlobFromReaderThatReturnsDataAndError(t *testing.T) {
	// See issue #264.
	// This tests the case of building a Blob from a reader who returns both data and an error for the final Read() call.
	assert := assert.New(t)
	tr := &testReader{buf: &bytes.Buffer{}}

	b := NewBlob(tr)

	actual := &bytes.Buffer{}
	io.Copy(actual, b.Reader())

	assert.True(bytes.Equal(actual.Bytes(), tr.buf.Bytes()))
	assert.Equal(byte(2), actual.Bytes()[len(actual.Bytes())-1])
}

File: js/src/blob-test.js
// @flow

import {blobType, refOfBlobType} from './type.js';
import {assert} from 'chai';
import Blob, {BlobReader, BlobWriter} from './blob.js';
import {suite, test} from 'mocha';
import {
  assertChunkCountAndType,
  assertValueHash,
  assertValueType,
  chunkDiffCount,
  testRoundTripAndValidate,
} from './test-util.js';
import {invariant} from './assert.js';
import {equals} from './compare.js';

// IMPORTANT: These tests and in particular the hash of the values should stay in sync with the
// corresponding tests in go

suite('Blob', () => {

  async function assertReadFull(expect: Uint8Array, reader: BlobReader): Promise<void> {
    const length = expect.length;
    let i = 0;
    let pos = reader._pos;

    while (i < length) {
      const next = await reader.read();
      assert.isFalse(next.done);
      const arr = next.value;
      invariant(arr);
      assert.strictEqual(arr.length + pos, reader._pos);
      pos = reader._pos;
      for (let j = 0; j < arr.length && i < length; j++) {
        assert.strictEqual(expect[i], arr[j]);
        i++;
      }
    }
  }

  async function testPrependChunkDiff(buff: Uint8Array, blob: Blob, expectCount: number):
      Promise<void> {
    const nb = new Uint8Array(buff.length + 1);
    for (let i = 0; i < buff.length; i++) {
      nb[i + 1] = buff[i];
    }

    const v2 = new Blob(nb);
    assert.strictEqual(expectCount, chunkDiffCount(blob, v2));
  }

  async function testAppendChunkDiff(buff: Uint8Array, blob: Blob, expectCount: number):
      Promise<void> {
    const nb = new Uint8Array(buff.length + 1);
    for (let i = 0; i < buff.length; i++) {
      nb[i] = buff[i];
    }

    const v2 = new Blob(nb);
    assert.strictEqual(expectCount, chunkDiffCount(blob, v2));
  }

  async function testRandomRead(buff: Uint8Array, blob: Blob): Promise<void> {
    const checkByteRange = async (start: number, rel: number, count: number) => {
      const buffSlice = new Uint8Array(buff.buffer, buff.byteOffset + rel + start, count);
      const blobReader = blob.getReader();
      assert.strictEqual(start, await blobReader.seek(start));
      assert.strictEqual(start, blobReader._pos);
      assert.strictEqual(start + rel, await blobReader.seek(rel, 1));
      assert.strictEqual(start + rel, blobReader._pos);
      await assertReadFull(buffSlice, blobReader);
    };

    const checkByteRangeFromEnd = async (length: number, offset: number, count: number) => {
      const buffSlice = new Uint8Array(buff.buffer,
                                       buff.byteOffset + buff.byteLength + offset,
                                       count);
      const blobReader = blob.getReader();
      assert.strictEqual(length + offset, await blobReader.seek(offset, 2));
      assert.strictEqual(length + offset, blobReader._pos);
      await assertReadFull(buffSlice, blobReader);
    };

    const length = buff.byteLength;
    let start = 0;
    let count = length / 2;
    while (count > 2) {
      await checkByteRange(start, 0, count);
      await checkByteRange(0, start, count);
      await checkByteRange(Math.floor(start / 2), Math.ceil(start / 2), count);
      await checkByteRangeFromEnd(length, start - length, count);
      start += count;
      count = (length - start) / 2;
    }
  }

  function randomBuff(len: number): Uint8Array {
    const r = new CountingByteReader();
    const a = new Uint8Array(len);
    for (let i = 0; i < len; i++) {
      a[i] = r.nextUint8();
    }
    return a;
  }

  async function blobTestSuite(size: number, expectHashStr: string, expectChunkCount: number,
                               expectPrependChunkDiff: number,
                               expectAppendChunkDiff: number) {
    const length = 1 << size;
    const buff = randomBuff(length);
    const blob = new Blob(buff);

    assertValueHash(expectHashStr, blob);
    assertValueType(blobType, blob);
    assert.strictEqual(length, blob.length);
    assertChunkCountAndType(expectChunkCount, refOfBlobType, blob);

    await testRoundTripAndValidate(blob, async(b2) => {
      await assertReadFull(buff, b2.getReader());
    });

    await testPrependChunkDiff(buff, blob, expectPrependChunkDiff);
    await testAppendChunkDiff(buff, blob, expectAppendChunkDiff);
    //console.profile('testRandomRead');
    await testRandomRead(buff, blob);
    //console.profileEnd('testRandomRead');
  }

  class CountingByteReader {
    _z: number;
    _value: number;
    _count: number;

    constructor(seed: number = 0) {
      this._z = seed;
      this._value = seed;
      this._count = 4;
    }

    nextUint8(): number {
      // Increment number
      if (this._count === 0) {
        this._z++;
        this._value = this._z;
        this._count = 4;
      }

      // Unshift a uint8 from our current number
      const retval = this._value & 0xff;
      this._value = this._value >>> 8;
      this._count--;

      return retval;
    }
  }

  test('Blob 1K', () =>
    blobTestSuite(10, 'sha1-f9fc78f387d90a334b85270a46484ebb86f32a3f', 3, 2, 2));

  test('LONG: Blob 4K', () =>
    blobTestSuite(12, 'sha1-060e57a95676be6078a2958f4586a8b3d3e6723d', 9, 2, 2));

  test('LONG: Blob 16K', () =>
    blobTestSuite(14, 'sha1-73d3ce5da681cf651509194ce63a055c51790385', 33, 2, 2));

  test('LONG: Blob 64K', () =>
    blobTestSuite(16, 'sha1-130b214a2e7edbdd5583dc7920b52838cad49471', 4, 2, 2));

  test('LONG: Blob 256K', () =>
    blobTestSuite(18, 'sha1-42d53b4f225322f70d725d53f8bc631d4549b6e4', 13, 2, 2));

  test('BlobWriter', () => {
    const a = randomBuff(15);
    const b1 = new Blob(a);
    const w = new BlobWriter();
    w.write(new Uint8Array(a.buffer, 0, 5));
    w.write(new Uint8Array(a.buffer, 5, 5));
    w.write(new Uint8Array(a.buffer, 10, 5));
    w.close();
    const b2 = w.blob;
    const b3 = w.blob;
    assert.strictEqual(b2, b3);
    assert.isTrue(equals(b1, b2));
  });

  test('BlobWriter close throws', () => {
    const a = randomBuff(15);
    const w = new BlobWriter();
    w.write(a);
    w.close();

    let ex;
    try {
      w.close();  // Cannot close twice.
    } catch (e) {
      ex = e;
    }
    assert.instanceOf(ex, TypeError);
  });

  test('BlobWriter write throws', () => {
    const a = randomBuff(15);
    const w = new BlobWriter();
    w.write(a);
    w.close();

    let ex;
    try {
      w.write(a);  // Cannot write after close.
    } catch (e) {
      ex = e;
    }
    assert.instanceOf(ex, TypeError);
  });
});
